test/cpp/cpp_test_util.cpp

```diff
@@ -304,12 +311,41 @@ void TestBackward(
   torch::Tensor output = testfn(input_vars);
   torch::Tensor xoutput = testfn(xinput_vars);
   AllClose(output, xoutput, rtol, atol);
-  output.backward(torch::ones_like(output));
-  xoutput.backward(torch::ones_like(xoutput));
-  for (size_t i = 0; i < inputs.size(); ++i) {
-    if (inputs[i].defined() && inputs[i].requires_grad()) {
-      ASSERT_TRUE(xinput_vars[i].grad().defined());
-      AllClose(input_vars[i].grad(), xinput_vars[i].grad(), rtol, atol);
+
+  // Calculating double backward requires create_graph=true
+  auto grads = torch::autograd::grad(
+      {output}, inputs_w_grad,
+      /*grad_outputs=*/{torch::ones_like(output)},
+      /*retain_graph=*/c10::nullopt, /*create_graph=*/double_backward);
+  auto xgrads = torch::autograd::grad(
+      {xoutput}, xinputs_w_grad,
+      /*grad_outputs=*/{torch::ones_like(xoutput)},
+      /*retain_graph=*/c10::nullopt, /*create_graph=*/double_backward);
+  for (size_t i = 0; i < grads.size(); ++i) {
+    ASSERT_TRUE(xgrads[i].defined());
+    AllClose(grads[i], xgrads[i], rtol, atol);
+  }
+
+  if (double_backward) {
+    // Check grad of sum(grads) w.r.t inputs_w_grad.
+    torch::Tensor sum = torch::zeros_like(grads[0]).sum();
+    torch::Tensor xsum = torch::zeros_like(xgrads[0]).sum();
+    for (int i = 0; i < grads.size(); ++i) {
+      if (!grads[i].requires_grad()) continue;
```

Comment:
One statement per line 😉 
But here instead of continuing (given that the indentation is small), I'd rather do the sums upon the inverse condition.